### ISSUE TITLE
Move javascript loading to bottom of page

### ISSUE THREAD

**Issue by phlax** [translate/pootle#4910]
Currently loading the javascript is blocking the page load. Moving the js resources to the end of the page should speed up rendering

**Comment by iafan** [comment]
I think this was already discussed before. In short, moving all JS to the bottom doesn't make the _application_ faster. Yes, it allows the browser to render some _static content_ earlier, before the scripts load, but the end result might look odd. This technique is good for mostly static pages, but in case of our app, which heavily relies on scripts, one might see semi-initialized page being rendered, which doesn't make sense as you can't consume the page content nor interact with the page until all scripts are loaded and run, anyway.

If one moves all the scripts to the bottom, they also need to ensure that they deliberately hide the uninitialized parts of the UI by providing default CSS styles. And once this is done, we are essentially where we were before: UI is not rendered until all scripts are loaded.

This is a fresh and still-to-be-solved example of the page content being rendered prematurely: https://github.com/translate/pootle/issues/4885

Having said that, there _is_ a reason to move all scripts to the bottom, though it is different from the one described in the case as I understood it: doing so, and also providing template and CSS defaults, might allow us to implement a "loader" UI so that the user sees something earlier. This is similar to the technique iOS apps use when they initially show a static "UI background" or "loading..." screenshot when opening the app and replace it with the real interactive UI a moment later.

I also think @julen is planning to work on reloading the browser UI without reloading the page, which will also improve the responsiveness of the UI.

**Comment by phlax** [comment]
@iafan i have quite a bit of experience doing this, and in many cases it does speed up page load, as it the page doesnt block while fetching the resources.

best practice is afaict (and **certainly** if you want to improve your page speed score as measured by google):

- load css inline
- load js at the end of the page

**Comment by phlax** [comment]
for reference here is the last site i worked on - which uses these techniques https://www.storenextdoor.com/

and here is the current page speed score for mlo https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fmozilla.locamotion.org

note the "should fix" section

**Comment by phlax** [comment]
related to this actually - /me wonders why we dont put the stats in the html in the first place

**Comment by iafan** [comment]
It's good to know that you have experience working on this. My primary point is that moving scripts to the bottom in our particular case likely needs to be combined with other changes so that we don't see semi-rendered pages and elements flashing and jumping around on page load. If we are on the same page on this, I'm game.

**Comment by iafan** [comment]
> related to this actually - /me wonders why we dont put the stats in the html in the first place

Me too. If we did, our browsing pages would be mostly static, and moving JS to the bottom would make much sense. However, this will mean we need to support browsing table rendering on both backend and frontend. Might be a good use case for server-side React.

**Comment by phlax** [comment]
@iafan its pretty simple tbh to do this with the current django views

**Comment by julen** [comment]
>> related to this actually - /me wonders why we dont put the stats in the html in the first place
> 
> Me too.

No need to wonder, [this is already done](
https://github.com/translate/pootle/blob/4fcaef1770f8d5ab5514429b5784cfb7370b79a0/pootle/templates/browser/index.html#L192
).

**Comment by phlax** [comment]
> No need to wonder, this is already done.

...in javascript

**Comment by julen** [comment]
I subscribe what @iafan said on [a previous comment](https://github.com/translate/pootle/issues/4910#issuecomment-233201009): this requires some other changes because of content flashing and related issues (automated page speed tests won't tell you about this).

On the other hand, I believe perceived performance is more important regardless of where scripts are put, so personally I'd put more attention on specific things affecting this, such as #4885.

**Comment by phlax** [comment]
ill stage a site with the suggested improvements, and then we can discuss

**Comment by julen** [comment]
>> No need to wonder, this is already done.
> 
> ...in javascript

I see what you mean, although frankly, this is moving in the direction to make the browsing table a component (plus reloadless browsing), so rendering it server side with Django templates won't help in the mid-term.

**Comment by phlax** [comment]
> I see what you mean, although frankly, this is moving in the direction to make the browsing table a component (plus reloadless browsing), so rendering it server side with Django templates won't help in the mid-term.

i think that requires some discussion. Personally i believe that the pages should be properly marked up with semantic content that SEs can index. Any js should be extra

**Comment by phlax** [comment]
@julen @iafan i have staged a branch with this change

The atf content loads much faster, and the page loads ~10% faster overall.

There is however a visible `fouc` or rather the reverse - some of the elements are initially missing.

Im not going to push this change for the current beta release, but i think its important that we sort this out fairly quickly

**Comment by iafan** [comment]
Am I right that this speeds up page loading by 10% __only for the first time__ user visits Pootle (because otherwise all JavaScript, CSS and images are expected to be cached forever and then loaded instantly)?

**Comment by phlax** [comment]
it certainly has less of an impact but still the page is rendered before loading all the javascript.

Also depending on what view you are clicking between - and if you are logging in etc, it still makes a significant difference

caching is not an excuse for poor performance

**Comment by iafan** [comment]
> it still makes a significant difference

How significant? Any numbers?

Where does the necessity for this optimization come from? Probably sharing some user feedback that we're not aware about might help adjusting our assessment of the problem, but from my perspective there's no poor performance here.

With the 10% number you provided this looks like a premature optimization to me — it offers minor load speed up in a statistical minority of page loads at the sake of rendering the page prematurely in an incorrect state.

**Comment by phlax** [comment]
> How significant? Any numbers?

no not really - only anecdotal im afraid

> With the 10% number you provided this looks like a premature optimization to me

im pretty shocked that you would say that. The ui experience on pootle is pretty poor

**Comment by phlax** [comment]
@iafan your argument seem to be that because we render the initial content badly, we should make the user wait.

Personally i want to make the page load as efficiently and with the best UX possible. Not simply to improve what is there - but so that we can build on what is there

**Comment by iafan** [comment]
> your argument seem to be that because we render the initial content badly, we should make the user wait

My argument is (and always was) that when it comes to performance, I'd like to see numbers and know how many users are affected and how often. When you say "it still makes a significant difference" but can't provide any measurable numbers this just means we should read this as "it still makes no measurable difference".

We shouldn't make the user wait, because we already don't make them wait on the resource loading part by itself. We definitely have room for optimization on data retrieval side before the page is served, on search speed — this is where I believe we could get the most impact.

If you consider this kind of load optimization a priority nevertheless, and you want to spend time addressing the initial content rendering problem, go for it. Again, please get me right here, **I would have no objections at all** if I'd be getting this speedup "for free" and with no side effects. But in your comments above you're essentially saying that you have made the change, noticed some subjective non-measurable speedups in edge cases, and now you ask Julen and me to spend time helping fix the introduced issues. This is not "for free" and this is why I think it is not worth it.

> The ui experience on pootle is pretty poor

I'd like to learn more about this. Probably in Gitter.

**Comment by unho** [comment]
There are two PRs attempting to achieve this:

- #4921
- #5513
